### PR TITLE
[ML-37954] Add notebook/job lineage when loading a model from UC

### PR DESCRIPTION
**Pull request by shichengzhou-db**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/shichengzhou-db/mlflow/pull/11305?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/11305/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 11305
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?
- During load_model, check if mlflow is running inside databricks notebook/job env, pass lineage info to UC artifact store
- In UC artifact store, when generating temporary credentials, pass the lineage info to UC
- Unit tests to check that load_model on a UC model uri will cause the api to generate temporary credentials to carry lineage header info

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.
Support emitting downstream model -> notebook/job lineage when loading a model in databricks environment

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [x] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by shichengzhou-db** [comment]
@mlflow-automation autoformat

File: tests/store/lineage/test_downstream_lineage.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        job_entity = Job(id=str(5678))
+        entity_list.append(Entity(job=job_entity))
+
+    expected_lineage_header_info = LineageHeaderInfo(entities=entity_list) if entity_list else None
+
+    with mock.patch(
+            "mlflow.utils.databricks_utils.get_config"
+    ), mock.patch(
+        "mlflow.utils.databricks_utils.is_in_databricks_notebook",
+        return_value=True,
```
Comment (kriscon-db) [inline tests/store/lineage/test_downstream_lineage.py:74]: can we add another test for when the is_in_databricks_(notebook|job) are both false assert that the lineage info is None?

**Comment by kriscon-db** [comment]
Mostly looks good to me.  Think we should add one more test to make sure lineage is None when the conditions are met.

File: tests/store/lineage/test_downstream_lineage.py (hunk 1 above)
Comment (shichengzhou-db) [reply tests/store/lineage/test_downstream_lineage.py]: thanks for reviewing. 
updated pr with test you suggested.

**Comment by kriscon-db** [comment]
LGTM from the code side.  Last thing....did we check this on a test shard?

**Comment by shichengzhou-db** [comment]
> LGTM from the code side. Last thing....did we check this on a test shard?

Yes, I did this on the UC change pr(see description of that pr), it works.

**Review by kriscon-db — APPROVED** [review]

File: mlflow/store/_unity_catalog/lineage/__init__.py
Comment (harupy) [inline mlflow/store/_unity_catalog/lineage/__init__.py:1]: Is this file necessary?

File: mlflow/pyfunc/__init__.py [hunk 2]
```diff
@@ -2249,14 +2279,10 @@ def predict(model_input: List[str]) -> List[str]:
...
             message=(
-                "The following sets of parameters cannot be specified together: {first_set_keys}"
-                " and {second_set_keys}. All parameters in one set must be `None`. Instead, found"
-                " the following values: {first_set_entries} and {second_set_entries}".format(
-                    first_set_keys=first_argument_set.keys(),
-                    second_set_keys=second_argument_set.keys(),
-                    first_set_entries=first_argument_set,
-                    second_set_entries=second_argument_set,
-                )
+                f"The following sets of parameters cannot be specified together:"
```
Comment (harupy) [inline mlflow/pyfunc/__init__.py:2282]: It looks like you're using a newer version of ruff. `0.2.1` should be used.

https://github.com/mlflow/mlflow/blob/c85ff32270a9879d6be9851505fadaee4824a1de/requirements/lint-requirements.txt#L1

File: mlflow/pyfunc/__init__.py [hunk 3]
```diff
@@ -861,7 +874,26 @@ def load_model(
...
+    if databricks_utils.is_in_databricks_notebook() or databricks_utils.is_in_databricks_job():
+        entity_list = []
+        # Get notebook id and job id, pack them into lineage_header_info
+        notebook_id = databricks_utils.get_notebook_id()
+        if notebook_id:
+            notebook_entity = Notebook(id=str(notebook_id))
+            entity_list.append(Entity(notebook=notebook_entity))
+
+        job_id = databricks_utils.get_job_id()
+        if job_id:
```
Comment (harupy) [inline mlflow/pyfunc/__init__.py:888]:
```suggestion
        if job_id := databricks_utils.get_job_id():
```

nit

File: mlflow/pyfunc/__init__.py [hunk 4]
```diff
@@ -861,7 +874,26 @@ def load_model(
...
                 release without warning.
     """
-    local_path = _download_artifact_from_uri(artifact_uri=model_uri, output_path=dst_path)
+
+    lineage_header_info = None
+    if databricks_utils.is_in_databricks_notebook() or databricks_utils.is_in_databricks_job():
+        entity_list = []
+        # Get notebook id and job id, pack them into lineage_header_info
+        notebook_id = databricks_utils.get_notebook_id()
+        if notebook_id:
```
Comment (harupy) [inline mlflow/pyfunc/__init__.py:883]:
```suggestion
        if notebook_id := databricks_utils.get_notebook_id():
```

File: mlflow/pyfunc/__init__.py [hunk 5]
```diff
@@ -861,7 +874,26 @@ def load_model(
...
+            notebook_entity = Notebook(id=str(notebook_id))
+            entity_list.append(Entity(notebook=notebook_entity))
+
+        job_id = databricks_utils.get_job_id()
+        if job_id:
+            job_entity = Job(id=str(job_id))
```
Comment (harupy) [inline mlflow/pyfunc/__init__.py:889]: Isn't `job_id` already a string?

Comment (shichengzhou-db) [reply mlflow/pyfunc/__init__.py:889]: ah i was copy-pasting code from another place. 
confirmed it's a string already

File: mlflow/store/_unity_catalog/lineage/__init__.py
Comment (shichengzhou-db) [reply mlflow/store/_unity_catalog/lineage/__init__.py:1]: not in particular, just copy pasting what every other module does, should I remove it?

File: mlflow/pyfunc/__init__.py (hunk 2 above)
Comment (shichengzhou-db) [reply mlflow/pyfunc/__init__.py:2282]: I remember this is some lint tests telling me these lines are too long and I manually fixed them to pass those tests.

**Comment by shichengzhou-db** [comment]
@mlflow-automation autoformat

File: tests/store/lineage/__init__.py
Comment (harupy) [inline tests/store/lineage/__init__.py:1]: Is this file necessary?

File: tests/store/lineage/test_downstream_lineage.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        mlflow.pyfunc.load_model(model_uri)
+        extra_headers = lineage_header_info_to_extra_headers(expected_lineage_header_info)
+        if is_in_notebook or is_in_job:
+            mock_http.assert_called_once_with(
+                host_creds=mock.ANY,
+                endpoint=mock.ANY,
+                method=mock.ANY,
+                json=mock.ANY,
+                extra_headers=extra_headers,
+            )
```
Comment (harupy) [inline tests/store/lineage/test_downstream_lineage.py:108]: Should `mock_http` be called with `extra_headers=None` if `is_in_notebook or is_in_job` is false?

**Review by harupy — APPROVED** [review]
Left a few minor comments, otherwise LGTM!

File: tests/store/lineage/test_downstream_lineage.py (hunk 6 above)
Comment (shichengzhou-db) [reply tests/store/lineage/test_downstream_lineage.py:108]: it should for how the code is currently written, but not super critical to test, the backend will treat {} / None as "there's no lineage info"

File: tests/store/lineage/__init__.py
Comment (shichengzhou-db) [reply tests/store/lineage/__init__.py:1]: will delete
